### DISCUSSION TITLE
[Feature] Ability to search for photos that have no location (not null island/0,0)

### DISCUSSION THREAD

**Discussion by frejaya** [immich-app/immich#8362] [Feature Request]
### The feature

I have a lot of photos taken with digital cameras that don't have GPS built in so I have to add the locations manually based on what I remember. I'd like to be able to filter my photos or search for photos that have no location so I can treat this like a to do list of photos that I still need to review. The photos don't have lat/long 0,0 (null island), they just have no location coordinates at all. I would prefer to do this right in the Immich interface and not through a database because even the dates aren't always right on the photos, the names of the files aren't useful, and I need to look at the picture itself to determine when and where it was actually taken. Some of the cameras used had issues where they would reset the date to some random starting date whenever they completely decharged.

I would also be fine with some kind of visual indicator on the photo itself, kind of like how you can toggle on/off the "storage indicator on asset tiles" on the mobile app. I have to go through photos carefully so if there was an icon for when they do have a location or when they don't have a location (either way) then I would be able to quickly at a glance see which ones I need to fix as I scroll through my timeline.

Edit to add: I know Digikam has this visual indicator and I do use it to update some coordinates but it's not as convenient as Immich. I'd prefer to do it within Immich itself. 

### Platform

- [ ] Server
- [X] Web
- [X] Mobile

**Comment by NIronwolf** [comment]
I'd love to see this added to search as well.

I'm picturing it right next to the likes of "Not in any album", "Archive", and "Favorite" under "DISPLAY OPTIONS".  Something like "No location".

Fortunately, the photos then come up in date order which puts them in a close group so I can at least get them shuffled around to the right city.

**Comment by foux** [comment]
Would be a huge improvment on search

**Comment by Cosgarion** [comment]
sure a good improvement.

**Comment by klendgit** [comment]
That would be a good improvement, I second that. For those who have SQL skills it is relatively easy to find those pictures. What I did was to add a port map from 5432 on the docker-compose.yaml file so the postgress DB is reachable from 'outside'. It is then easy to find those pictures and their filepath.
Still, it is much more convenient for everyone to be able to search for material **without** lattitude or longitude values.
Second to this request, I think it would be very convenient to be able to multi-select pictures and give them the same location 'in one go' - that is currently not possible via filtering.

**Reply by NIronwolf** [reply]
> Second to this request, I think it would be very convenient to be able to multi-select pictures and give them the same location 'in one go' - that is currently not possible via filtering.

As a workaround if they don't add this, you can put them in a temporary album, and on the album screen you are able to bulk change location.

**Reply by klendgit** [reply]
Thanks, that is a good suggestion - I did not know that!

**Reply by jetpacktuxedo** [reply]
There is bulk location setting now directly in the image viewer window (but still no way to search for photos with no location metadata)

**Comment by jforseth210** [comment]
What if there was an "Unknown" option in the location selectors? The frontend could pass `country: null` to the server, and, assuming that `BaseSearchDto` was updated to allow this, I _think_ `searchAssetBuilder` would look for assets with no country in the  `exifInfo` just like that. 

I don't really understand how the reverse geocoding works, but there'd probably be a few cases where the photos have a location but no country, like in the middle of the ocean or something, but I think it'd work pretty well.

**Comment by johnmmcgee** [comment]
This would be a great addition.

**Comment by steveredden** [comment]
+1 for request :D

@klendgit do you mind sharing your SQL query (and/or insert to create an album) you used?  I'm just getting started and haven't fully digested the db schema...

**Reply by klendgit** [reply]
I will try ;-) - but I did not create an album, but used sql statements to add location and cities based on the **dateTimeOriginal**.

Firstly, I opened up the database port to the outside by adding 
```
    ports:
      - 6543:5432
```
 to the database part of **docker-compose.yml**
 This makes it possible to reach the DB on port 6543 from the outside, with, for instance, DBeaver.
 Connection to it works via the variables defined in the file **.env**
 Once you are able to connect to the DB, I used this query to get the images that do not have an exif geo value:
```
select * from exif e inner join assets a on e."assetId" = a.id where e.latitude is null
```
Note: this returns any image without location, also Whatsapp images, etc, so you might want to narrow down with additional clauses like: `and a."originalFileName" not like '%-WA'` or limit via `e.make` or `e.model`

However, when you have a select query that returns all your images without a location you could assign a location to them with **update** - I updated my images with a location in the country **Mali** in Africa, where I was sure that I did not have images from.
See this example:
```
update exif set latitude = 20.598993432651632, longitude = -3.547410987535704 where exif."assetId" in (
 select e."assetId" from exif e inner join assets a on e."assetId" = a.id
where e.latitude is null and a."originalFileName" not like '%-WA%' and a."originalPath" like '/usr/src%'
)
```
The sql above limits the updating of images to only certain digital images that do not have a filename with **-WA** in them and were not uploaded but in my library. This can be different for every installation. **Use sql update statements with care**

The final step I made was to update those images that have a location latitude = 20.598993432651632, longitude = -3.547410987535704 to update the **_exif.latitude_** to the **dateTimeOrignal**, the **_exif.country_** to '**Mali**' and the **_exif.city_** to the **date pattern of the dateTimeOriginal**. I did this by an export of the exif information and using some linux tools. I'm sorry but I do not have that information anymore.
The trick was to retrieve the exif.assetId,exif.dateTimeOriginal for those images into a csv file on linux. I made sql statements from it where the '2018-08-06 20:51:54.000 +0200' was converted to a latitude of **20.180806** and the city was converted to '**2018-08-06**' - I am sorry but I do not have an example anymore of that process. An example of an update sql statement is this:
`update exif set city = '2018-08-06', country = 'Mali', latitude = 20.180806 where assetId = '593f22ec-1f6f-4fa0-b5ec-a7ea170484ee'`

In the end, I had a list of images that were all in a certain longitude (-3.54 in my case) and were all on a latitude **based on the day the picture was shot.** I then was able to use the normal tools to quick select those images where the city was 2018-08-06 and give them the correct location. (Because typically, images shot on one certain day were taken in a relative similar location)

I hope this helps - **please use extreme care when you are dealing with the immich DB** and if you are uncertain, I suggest to not do this exercise. It helped me a lot but I am familiar with sql and scripts on linux.

**Reply by steveredden** [reply]
Thanks for the very! in depth overview!

**Reply by steveredden** [reply]
I ended up with the following, performed in pgadmin against the database:

1. Manually create an empty album
1. Identify its `id`: `select * from albums where albumName = 'Locationless'`
1. Insert assets with `null` latitude into the album, with hardcoded `albumsId`
   ```
   INSERT INTO albums_assets_assets ("albumsId", "assetsId")
   select '4e21c4c8-cd76-4db0-b2c2-93046f724ef9', e."assetId"
   from exif e
   inner join assets a on e."assetId" = a.id
   where e.latitude is null
   ```

(asset album)
![image](https://github.com/user-attachments/assets/add0de4f-d65e-40f8-9407-237c72ac3693)

my plan is to assign a real location as time permits, then remove the assets with a location from the album.

Turns out once assigning a location, the collection is 'unchecked' so I'm now removing them via sql too:

```
DELETE FROM albums_assets_assets
WHERE "assetsId" IN (
	SELECT e."assetId"
	FROM exif e
	INNER JOIN assets a ON e."assetId" = a.id
	WHERE e.latitude IS NOT NULL
);
```

(after updating location of ~491 assets)
![image](https://github.com/user-attachments/assets/176750d4-9e50-44e5-8879-55b0a6443473)

**Reply by makuhlmann** [reply]
> I ended up with the following, performed in pgadmin against the database:
> 
> 1. Manually create an empty album
> 2. Identify its `id`: `select * from albums where albumName = 'Locationless'`
> 3. Insert assets with `null` latitude into the album, with hardcoded `albumsId`
>    ```
>    INSERT INTO albums_assets_assets ("albumsId", "assetsId")
>    select '4e21c4c8-cd76-4db0-b2c2-93046f724ef9', e."assetId"
>    from exif e
>    inner join assets a on e."assetId" = a.id
>    where e.latitude is null
>    ```
> 

Small side note: if you have multiple user accounts, the following will make sure that only the images assigned to the owner of the album will be added to it. Be aware that it may break if there are multiple albums with the same name.

```
INSERT INTO albums_assets_assets ("albumsId", "assetsId")
select (select id from albums where albums."albumName" = 'Locationless'), e."assetId" 
from exif e
inner join assets a on e."assetId" = a."id"
where e.latitude is null and a."ownerId" = (select "ownerId" from albums where albums."albumName" = 'Locationless')
```

---

> my plan is to assign a real location as time permits, then remove the assets with a location from the album.
> 
> Turns out once assigning a location, the collection is 'unchecked' so I'm now removing them via sql too:
> 
> ```
> DELETE FROM albums_assets_assets
> WHERE "assetsId" IN (
> 	SELECT e."assetId"
> 	FROM exif e
> 	INNER JOIN assets a ON e."assetId" = a.id
> 	WHERE e.latitude IS NOT NULL
> );
> ```

**That SQL query is dangerous! Do not run it!** It will remove assets from **ANY** album if the location has been set. Here is a more appropriate query filtering for the album. Be aware that it may break if there are multiple albums with the same name.

```
delete from albums_assets_assets
WHERE "assetsId" IN (
	SELECT e."assetId"
	FROM exif e
	INNER JOIN assets a ON e."assetId" = a.id
	WHERE e.latitude IS NOT NULL
) and "albumsId" = (select id from albums where albums."albumName" = 'Locationless');
```

**Comment by jforseth210** [comment]
Also #10866 was merged. You can go to country and select "Unknown" in the latest version.

**Comment by Kugeleis** [comment]
A search/filter option inversion for any property would be great to consider. 
e. g. person foo not in image
not tag `foo`

**Reply by jforseth210** [reply]
Out of scope for this feature request, you should open a new one if there isn’t one already.

**Reply by Kugeleis** [reply]
I intended to lift this feature request to a more general level.
Here: Filter for photos with **no** location.
General: Filter for photos with **not** *a property*.

**Comment by duncaninnes** [comment]
This would be excellent feature for improving the location data of my library.  Also for weeding out the GIFs, WhatsApp, and collage's which have been put in there.  I don't want to delete them all.  There are some important WhatsApp images in there which require a location added to them.

**Reply by miasinterestinglife** [reply]
There is now the option to filter by location on the main page. You can just select the country as unknown and you get all the pictures without a location. For ease of use you might want to scroll all the way to the bottom and add all to an album (depending on how many it is, you might have to do this multiple times as there appears to be a limit on how many assets can be added to an album at once)

**Reply by duncaninnes** [reply]
This doesn't quite provide the same functionality though.  Country = Unknown also matches many pictures I have taken at sea.  These images all have accurate location data, but the calculation for 'country' clearly doesn't cope with them being offshore.  Some of these pictures are only minimally off the coast.  I'm looking at one now where we are less than a mile from the shore of the Cambodian island of Koh Rong.

**Reply by miasinterestinglife** [reply]
Ah that makes sense, it's just never happened to me so far.
